### PR TITLE
Add new Kanister function WaitV2

### LINKED ISSUE #1583: Wait function condition arg should only use go-template an not a mix of jsonpath and go-template
Today this usage of the [Wait](https://docs.kanister.io/functions.html#wait) function that seems natural does not work
```
- func: Wait
      name: waitCouchbaseClusterReadyAndBalanced
      args:
        timeout: 240s
        conditions:
          allOf:
            - condition: '{{ $available := false }}{{ range $condition := $.status.conditions }}{{ if and (eq .reason "Available") (eq .status "True")  }}{{ $available = true }}{{ end }}{{ end }}{{ $available }}'
              objectReference:
                apiVersion: couchbase.com/v2
                resource: couchbaseclusters
                name: "{{ .Object.metadata.name }}"
                namespace: "{{ .Object.metadata.namespace}}"
            - condition: '{{ $available := false }}{{ range $condition := $.status.conditions }}{{ if and (eq .reason "Balanced") (eq .status "True")  }}{{ $available = true }}{{ end }}{{ end }}{{ $available }}'
              objectReference:
                apiVersion: couchbase.com/v2
                resource: couchbaseclusters
                name: "{{ .Object.metadata.name }}"
                namespace: "{{ .Object.metadata.namespace}}"
```

Because the template has different level of rendering and at some point expression like `$.status.conditions` are evaluated as jsonpath. This is confusing and should be simplified.

**Describe the solution you'd like**
if `kubecl get -o go-template ...` evaluate as true for instance 
```
k get couchbaseclusters.couchbase.com cb-example -o go-template='{{ $available := false }}{{ range $condition := $.status.conditions }}{{ if and (eq .reason "Available") (eq .status "True")  }}{{ $available = true }}{{ end }}{{ end }}{{ $available }}'
true
```

Then the conditions  
```
            - condition: '{{ $available := false }}{{ range $condition := $.status.conditions }}{{ if and (eq .reason "Available") (eq .status "True")  }}{{ $available = true }}{{ end }}{{ end }}{{ $available }}'
```

should render and evaluate to true also.

**Describe alternatives you've considered**
For the moment the workaround we found is this construct 
```
            - condition: '{{ if (eq "{$.status.conditions[?(@.reason=="Available")].status}" "True") }}true{{ else }}false{{ end }}'            
              objectReference:
                apiVersion: v2
                group: couchbase.com
                resource: couchbaseclusters
                name: "{{ .Object.metadata.name }}"
                namespace: "{{ .Object.metadata.namespace}}"   
```

But this part `$.status.conditions[?(@.reason=="Available")].status` is not supported by go-template because it's a jsonpath, hence it's impossible to test your expression before you launch the actionset. There is no template engine that work with this mix.

One have to know when go-template stop and when jsonpath start it's pretty tricky.

**Environment**
Kubernetes Version/Provider: 1.22
Storage Provider: azure
Cluster Size (#nodes): 3
Data Size: 20Gb

**Additional context**
Kanister version 0.80.0

### PR DESCRIPTION
**Pull request by PrasadG193**
## Change Overview

This PR adds new function `WaitV2` improved version of existing `Wait` function which simplifies the condition definition. 
Instead of Jsonpath to fetch K8s object value, we can use go-template syntax which can be validated with kubectl.

e.g To check if the deployment is ready, once can form following go-template which can be validated with `kubectl` -

```
$ kubectl get deploy -n kanisterwaittest-b58j2 test-deployment-tntfb -o go-template='{{ $available := false }}{{ range $condition := $.status.conditions }}{{ if and (eq .type "Available") (eq .status "True")  }}{{ $available = true }}{{ end }}{{ end }}{{ $available }}' 
true
```
The same go-template can be used in the `WaitV2` function arg to define condition 

```
apiVersion: cr.kanister.io/v1alpha1
kind: Blueprint
metadata:
  name: wait-blueprint
actions:
  backup:
    phases:
    - func: WaitV2
      name: waitForRestoreReady
      args:
        timeout: 5m
        conditions:
          anyOf:
          - condition: '{{ $available := false }}{{ range $condition := $.status.conditions }}{{ if and (eq .type "Available") (eq .status "True") }}{{ $available = true }}{{ end }}{{ end }}{{ $available }}'
            objectReference:
              apiVersion: "v1"
              group: "apps"
              name: "{{ .Object.metadata.name }}"
              namespace: "{{ .Object.metadata.namespace }}"
              resource: "deployments"

```

## Pull request type

Please check the type of change your PR introduces:
- [ ] :construction: Work in Progress
- [ ] :rainbow: Refactoring (no functional changes, no api changes)
- [ ] :hamster: Trivial/Minor
- [ ] :bug: Bugfix
- [ ] :sunflower: Feature
- [ ] :world_map: Documentation
- [ ] :robot: Test

## Issues 

- fixes #1583

## Test Plan

- [ ] :muscle: Manual
- [ ] :zap: Unit test
- [ ] :green_heart: E2E

- Added unit tests
- Manually validated the BP mentioned in the PR description on a deployment

### CODE REVIEW

**Comment by PrasadG193** [comment]
image

**Review by michaelcourcy — APPROVED** [review]
LGTM, I have a small question in the code but I think I auto answer it, otherwise thanks for asking me the review the unit test you wrote were very interesting, the documentation is also very good.

File: pkg/function/waitv2.go [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+var _ kanister.Func = (*waitV2Func)(nil)
+
+type waitV2Func struct{}
+
+func (*waitV2Func) Name() string {
+	return WaitV2FuncName
+}
+
+func (ktf *waitV2Func) Exec(ctx context.Context, tp param.TemplateParams, args map[string]interface{}) (map[string]interface{}, error) {
```
[the remark this answers is not in the document]
Comment (pavannd1) [reply pkg/function/waitv2.go:52]: That's right - the Func interface expects the `Exec` to have `output, err` as return args.

**Comment by PrasadG193** [comment]
Rebased and fixed conflicts. @pavannd1 could you please take a look?

**Review by pavannd1 — APPROVED** [review]

**Pull request merged**